File: frontend/utils/chunker/token/splitBySentenceBoundaries.ts
import { countTokens } from "./countTokens";
import { splitByWords } from "./splitByWords";

/**
 * Split a single line by sentence boundaries, falling back to word split.
 * Ensures no trailing text is silently dropped.
 */
function splitBySentenceBoundaries(line: string, maxTokens: number): string[] {
  const sentenceMatch = line.match(/[^\.!\?]+[\.!\?]+|.+$/g);
  if (!sentenceMatch) return splitByWords(line, maxTokens);

  const parts: string[] = [];
  let current = "";
  for (const sentence of sentenceMatch) {
    const trimmed = sentence.trim();
    if (!trimmed) continue;
    const candidate = current ? current + " " + trimmed : trimmed;
    if (countTokens(candidate) <= maxTokens) {
      current = candidate;
    } else {
      if (current) parts.push(current.trim());
      if (countTokens(trimmed) > maxTokens) {
        parts.push(...splitByWords(trimmed, maxTokens));
        current = "";
      } else {
        current = trimmed;
      }
    }
  }
  if (current) parts.push(current.trim());
  return parts.length > 0 ? parts : splitByWords(line, maxTokens);
}

export { splitBySentenceBoundaries };

File: src-tauri/src/chunker/token/split_by_sentence_boundaries.rs
use regex::Regex;
use crate::chunker::token::count_tokens;
use crate::chunker::token::split_by_words;
use std::sync::LazyLock;

static RE_SENTENCE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"[^\.!\?]+[\.!\?]+|.+$").unwrap());

pub fn split_by_sentence_boundaries(text: &str, max_tokens: usize) -> Vec<(String, usize)> {
    let sentences: Vec<&str> = RE_SENTENCE.find_iter(text).map(|m| m.as_str()).collect();
    if sentences.is_empty() {
        return split_by_words(text, max_tokens);
    }

    let mut parts = Vec::new();
    let mut current = String::new();
    let mut current_tokens = 0;

    for sentence in sentences {
        let trimmed = sentence.trim();
        if trimmed.is_empty() {
            continue;
        }
        
        let sentence_tokens = count_tokens(trimmed);
        let separator_tokens = if current.is_empty() { 0 } else { 1 };

        if current.is_empty() {
            if sentence_tokens <= max_tokens {
                current = trimmed.to_string();
                current_tokens = sentence_tokens;
            } else {
                parts.extend(split_by_words(trimmed, max_tokens));
            }
        } else if current_tokens + separator_tokens + sentence_tokens <= max_tokens {
            current = format!("{} {}", current, trimmed);
            current_tokens += separator_tokens + sentence_tokens;
        } else {
            parts.push((current.trim().to_string(), current_tokens));
            if sentence_tokens > max_tokens {
                parts.extend(split_by_words(trimmed, max_tokens));
                current = String::new();
                current_tokens = 0;
            } else {
                current = trimmed.to_string();
                current_tokens = sentence_tokens;
            }
        }
    }

    if !current.is_empty() {
        parts.push((current.trim().to_string(), current_tokens));
    }

    if parts.is_empty() {
        split_by_words(text, max_tokens)
    } else {
        parts
    }
}
